### PR TITLE
filter:manager change metadata and downstream APIs.

### PR DESCRIPTION
**Pull request by alyssawilk**
Signed-off-by: Alyssa Wilk <alyssar@chromium.org>
This PR does two things.  

1) it moves the translation of metadata map types to the HCM, so we consistently use MetadataMapPtr&& everywhere in between the upstream codec and the downstream codec and
2) start moving downstream-only code out of the generic FM base class into a downstream-specific base class.  Specifically I'm claiming (and documenting) that if an upstream filter sends a local reply, as the response is not generated by the cluster so should only traverse the downstream filter chain, so UpstreamFM sendLocal Reply will just forward sendLocalReply to the DownstreamFM and HCM. 

Risk Level: low (minor refactors)
Testing: n/a
Docs Changes: n/a
Release Notes: n/a
Part of https://github.com/envoyproxy/envoy/issues/10455

### CODE REVIEW

**Comment by alyssawilk** [comment]
ping?

File: envoy/stream_info/stream_info.h [hunk 1]
```diff
@@ -728,6 +728,8 @@ class StreamInfo {
     downstream_info.setUpstreamBytesMeter(upstream_info.getUpstreamBytesMeter());
     upstream_info.setDownstreamBytesMeter(downstream_info.getDownstreamBytesMeter());
   }
+
+  virtual void dumpState(std::ostream& os, int indent_level = 0) const PURE;
```
Comment (snowp) [inline envoy/stream_info/stream_info.h:732]: doc string?

**Review by snowp — CHANGES_REQUESTED** [review]
Makes sense to me, just one comment

**Review by snowp — APPROVED** [review]
Thanks!

**Pull request merged**